cli/azd/pkg/tools/github/github.go

```diff
@@ -323,3 +377,185 @@ func isEnvVarTokenSource(message string) bool {
+	tarReader := tar.NewReader(gzReader)
+	for {
+		fileHeader, err := tarReader.Next()
+		if errors.Is(err, io.EOF) {
+			return extractedAt, fmt.Errorf("did not find gh cli within tar file: %w", err)
+		}
+		if fileHeader == nil {
+			continue
+		}
+		if err != nil {
+			return extractedAt, err
+		}
+		// Tha name contains the path, remove it
+		fileNameParts := strings.Split(fileHeader.Name, "/")
+		fileName := fileNameParts[len(fileNameParts)-1]
+		if fileHeader.Typeflag == tar.TypeReg && fileName == "gh" {
+			filePath := filepath.Join(dst, fileName)
+			ghCliFile, err := os.OpenFile(filePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, os.FileMode(fileHeader.Mode))
+			if err != nil {
+				return extractedAt, err
+			}
+			defer ghCliFile.Close()
+			_, err = io.Copy(ghCliFile, tarReader)
+			if err != nil {
+				return extractedAt, err
+			}
+			extractedAt = filePath
+			break
+		}
+	}
+
+	return extractedAt, nil
+}
+
+// extractGhCli gets the Github cli from either a zip or a tar.gz
+func extractGhCli(src, dst string) (string, error) {
+	if strings.HasSuffix(src, ".zip") {
+		return extractFromZip(src, dst)
+	} else if strings.HasSuffix(src, ".tar.gz") {
+		return extractFromTar(src, dst)
+	}
+	return "nil", fmt.Errorf("Unknown format")
+}
+
+// GetGitHubCliImplementation defines the contract function to acquire the GitHub cli.
+// The `outputPath` is the destination where the github cli is place it.
+type GetGitHubCliImplementation func(
+	ctx context.Context,
+	transporter policy.Transporter,
+	ghVersion semver.Version,
+	extractImplementation ExtractGitHubCliFromFileImplementation,
+	outputPath string) error
+
+// ExtractGitHubCliFromFileImplementation defines how the cli is extracted
+type ExtractGitHubCliFromFileImplementation func(src, dst string) (string, error)
+
+// downloadGh downloads a given version of GitHub cli from the release site.
+func downloadGh(
+	ctx context.Context,
+	transporter policy.Transporter,
+	ghVersion semver.Version,
+	extractImplementation ExtractGitHubCliFromFileImplementation,
+	path string) error {
+
+	// arm and x86 not supported (similar to bicep)
+	var releaseName string
+	switch runtime.GOOS {
+	case "windows":
+		releaseName = "gh_2.22.1_windows_amd64.zip"
+	case "darwin":
+		releaseName = "gh_2.22.1_macOS_amd64.tar.gz"
+	case "linux":
+		releaseName = "gh_2.22.1_linux_amd64.tar.gz"
+	default:
+		return fmt.Errorf("unsupported platform")
+	}
+
+	//https://github.com/cli/cli/releases/download/v2.22.1/gh_2.22.1_linux_arm64.rpm
+	ghReleaseUrl := fmt.Sprintf("https://github.com/cli/cli/releases/download/v%s/%s", ghVersion, releaseName)
+
+	log.Printf("downloading github cli release %s -> %s", ghReleaseUrl, releaseName)
+
+	spanCtx, span := telemetry.GetTracer().Start(ctx, events.GitHubCliInstallEvent)
+	defer span.End()
+
+	req, err := http.NewRequestWithContext(spanCtx, "GET", ghReleaseUrl, nil)
+	if err != nil {
+		return err
+	}
+
+	resp, err := transporter.Do(req)
+	if err != nil {
+		return err
+	}
```

Comment:
Can we add a retry here for any service blips (simply add it inline)? I think ultimately we need all components to consume a `transporter` that implements retry, likely built using the pipeline capabilities of azure sdk